archive/tar_test.go

```diff
@@ -698,6 +698,44 @@ func TestDiffTar(t *testing.T) {
 				fstest.CreateFile("/d2/f", []byte("ok"), 0644),
 			),
 		},
+		{
+			name: "HardlinkParentInclusion",
+			validators: []tarEntryValidator{
+				dirEntry("d2/", 0755),
+				fileEntry("d2/l1", []byte("link me"), 0644),
+				// d1/f1 and its parent is included after the new link,
+				// before the new link was included, these files would
+				// not habe needed
+				dirEntry("d1/", 0755),
+				linkEntry("d1/f1", "d2/l1"),
+				dirEntry("d3/", 0755),
+				fileEntry("d3/l1", []byte("link me"), 0644),
+				dirEntry("d4/", 0755),
+				linkEntry("d4/f1", "d3/l1"),
```

Comment:
no test for `d5`?

Reply:
This is the complete list of entries which should be in the tar, meaning if there is anything different, less or more than what is in this array, it will fail. This is testing that `d5` is not in this list, since `d5` does not have a change, but a link associated with the file in `d5` changed. Therefore `d5` is actually being tested, but just that the directory inclusion logic does not include the parent directory of a link source which has been removed.